singularity-eos/base/fast-math/logs.hpp

```diff
@@ -78,102 +62,216 @@
-#else
+double pow2_o1_aliased(const double x) {
+  using namespace FP64LE;
+  return as_double(static_cast<std::int64_t>(x * scale_up) + one_as_int);
+}
+// ----------------------------------------------------------------------
+
+// Second-order interpolation based NQTs
+// These implementations are due to Peter Hammond
+// ----------------------------------------------------------------------
+// Portable versions that use frexp/ldexp rather than integer aliasing
+PORTABLE_FORCEINLINE_FUNCTION
+double lg_o2_portable(const double x) {
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
+  constexpr double four_thirds = 4. / 3.;
+  int e;
+  const double m = frexp(x, &e);
+  return e - four_thirds * (m - 2) * (m - 1);
+}
+
+// This version uses the exact formula
+PORTABLE_FORCEINLINE_FUNCTION
+double pow2_o2_portable(const double x) {
+  // log2(mantissa). should go between -1 and 0
   const int flr = std::floor(x);
-  const double remainder = x - flr;
-  const double mantissa = 0.5 * (remainder + 1);
+  const double lm = x - flr - 1;
+  const double mantissa = 0.5 * (3 - std::sqrt(1 - 3 * lm));
   const double exponent = flr + 1;
-#ifndef SINGULARITY_USE_SINGLE_LOGS
-  // double precision is default
   return ldexp(mantissa, exponent);
-#else
-  // Faster but less accurate
-  return ldexpf((float)mantissa, exponent);
-#endif // SINGULARITY_USE_SINGLE_LOGS
-#endif // SINGULARITY_USE_HIGH_RISK_MATH
 }
 
+// Integer aliased/bithacked versions
 PORTABLE_FORCEINLINE_FUNCTION
-double lg(const double x) {
+double lg_o2_aliased(const double x) {
+  using namespace FP64LE;
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
+  const std::int64_t x_as_int = as_int(x) - one_as_int;
+  const std::int64_t frac_as_int = x_as_int & mantissa_mask;
+  const std::int64_t frac_high = frac_as_int >> 26;
+  const std::int64_t frac_low = frac_as_int & low_mask;
+  const std::int64_t frac_squared =
+      frac_high * frac_high + ((frac_high * frac_low) >> 25);
+
+  return static_cast<double>(x_as_int + ((frac_as_int - frac_squared) / 3)) * scale_down;
+}
+
+PORTABLE_FORCEINLINE_FUNCTION
+double pow2_o2_aliased(const double x) {
+  using namespace FP64LE;
+  constexpr std::int64_t a = 9007199254740992;  // 2 * 2^52
+  constexpr double b = 67108864;                // 2^26
+  constexpr std::int64_t c = 18014398509481984; // 4 * 2^52
+  const std::int64_t x_as_int = static_cast<std::int64_t>(x * scale_up);
+  const std::int64_t frac_as_int = x_as_int & mantissa_mask;
+  const std::int64_t frac_sqrt =
+      static_cast<std::int64_t>(b * std::sqrt(static_cast<double>(c - 3 * frac_as_int)));
 
-  assert(x > 0 && "log divergent for x <= 0");
+  return as_double(x_as_int + a - frac_sqrt - frac_as_int + one_as_int);
+}
+// ----------------------------------------------------------------------
 
-#ifndef SINGULARITY_USE_TRUE_LOG_GRIDDING
-  // Default expression
-  return fastlg(x);
+PORTABLE_FORCEINLINE_FUNCTION
+double fastlg(const double x) {
```

Comment:
So I was incorrectly expecting to see `sesame2spiner` to be touched by this MR... but I suspect it is not because all the logic for which log we use is handled by these `#ifdef`s.  Should `sesame2spiner` tables we marked with a header denoting which version of log they use (maybe they already are)?

Reply:
they aren't but they should be. I will add.